Starred expressions in subscriptions and `TypeVarTuple` not available in Python 3.10

### What happened?

When trying to run the `test_data_dmri.py` test suite, Python raises an `ImportError` because it does not like the 
```
class BaseDataset(Generic[*Ts]):
```

syntax.

I am using Python 3.10. According to my IDE:
```
Python version 3.10 does not support starred expressions in subscriptions
```

After removing the problematic `Generic[*Ts]` inheritance, the error is translated to other locations where the `[*Ts]` syntax is used, and once all removed I get
```
ImportError: cannot import name 'TypeVarTuple' from 'typing' (/usr/lib/python3.10/typing.py
```

See the additional information.

These are not being caught because even if we only require [Python >= 3.10](https://github.com/nipreps/nifreeze/blob/7237ecff29902c3e077e06303dc94c0e34ec1ef7/pyproject.toml#L21) , we are only testing with [Python 3.12](https://github.com/nipreps/nifreeze/blob/7237ecff29902c3e077e06303dc94c0e34ec1ef7/.github/workflows/test.yml#L107).

### What command did you use?

```shell
pytest test_data_dmri.py
```

### What version of the software are you running?

`HEAD` at commit 6697c7872cf77e3dd35288118f0294d80d6e7f48

### How are you running this software?

Local installation ("bare-metal")

### Is your data BIDS valid?

Yes

### Are you reusing any previously computed results?

No

### Please copy and paste any relevant log output.

```shell
ImportError while loading conftest 'src/nifreeze/test/conftest.py'.
test/conftest.py:33: in <module>
    from nifreeze.data.dmri import DWI
src/nifreeze/data/dmri.py:39: in <module>
    from nifreeze.data.base import BaseDataset, _cmp, _data_repr
E     File "src/nifreeze/src/nifreeze/data/base.py", line 61
E       class BaseDataset(Generic[*Ts]):
E                         ^^^^^^^^^^^^
E   SyntaxError: invalid syntax. Perhaps you forgot a comma?
```

### Additional information / screenshots

Starred expressions in subscriptions: https://bugs.python.org/issue45586
TypeVarTuple: https://peps.python.org/pep-0646/

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -85,6 +85,9 @@ jobs:
     # Unit tests only on Linux/Python 3.12
     runs-on: 'ubuntu-latest'
     needs: ['cache-test-data']
+    strategy:
+      matrix:
+        python-version: ['3.10', '3.11', '3.12']
     steps:
       - uses: actions/checkout@v4
         with:
@@ -93,7 +96,7 @@ jobs:
       - name: Set up Python ${{ matrix.python-version }}
         uses: actions/setup-python@v5
         with:
-          python-version: '3.12'
+          python-version: ${{ matrix.python-version }}
       - name: Install the latest version of uv
         uses: astral-sh/setup-uv@v5
       - name: Install ANTs

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -30,6 +30,7 @@ dependencies = [
     "scikit-image>=0.15.0",
     "scikit_learn>=1.3.0",
     "scipy>=1.8.0",
+    "typing_extensions >=4.12",
 ]
 dynamic = ["version"]
 

--- a/src/nifreeze/data/base.py
+++ b/src/nifreeze/data/base.py
@@ -27,7 +27,7 @@ from __future__ import annotations
 from collections import namedtuple
 from pathlib import Path
 from tempfile import mkdtemp
-from typing import Any, Generic, TypeVarTuple
+from typing import Any, Generic
 
 import attr
 import h5py
@@ -35,6 +35,7 @@ import nibabel as nb
 import numpy as np
 from nibabel.spatialimages import SpatialHeader, SpatialImage
 from nitransforms.linear import Affine
+from typing_extensions import TypeVarTuple, Unpack
 
 from nifreeze.utils.ndimage import load_api
 
@@ -58,7 +59,7 @@ def _cmp(lh: Any, rh: Any) -> bool:
 
 
 @attr.s(slots=True)
-class BaseDataset(Generic[*Ts]):
+class BaseDataset(Generic[Unpack[Ts]]):
     """
     Base dataset representation structure.
 
@@ -99,13 +100,12 @@ class BaseDataset(Generic[*Ts]):
 
         return self.dataobj.shape[-1]
 
-    def _getextra(self, idx: int | slice | tuple | np.ndarray) -> tuple[*Ts]:
-        # PY312: Default values for TypeVarTuples are not yet supported
+    def _getextra(self, idx: int | slice | tuple | np.ndarray) -> tuple[Unpack[Ts]]:
         return ()  # type: ignore[return-value]
 
     def __getitem__(
         self, idx: int | slice | tuple | np.ndarray
-    ) -> tuple[np.ndarray, np.ndarray | None, *Ts]:
+    ) -> tuple[np.ndarray, np.ndarray | None, Unpack[Ts]]:
         """
         Returns volume(s) and corresponding affine(s) through fancy indexing.
 

--- a/src/nifreeze/estimator.py
+++ b/src/nifreeze/estimator.py
@@ -26,9 +26,10 @@ from __future__ import annotations
 
 from pathlib import Path
 from tempfile import TemporaryDirectory
-from typing import Self, TypeVar
+from typing import TypeVar
 
 from tqdm import tqdm
+from typing_extensions import Self
 
 from nifreeze.data.base import BaseDataset
 from nifreeze.model.base import BaseModel, ModelFactory

--- a/test/test_data_base.py
+++ b/test/test_data_base.py
@@ -53,25 +53,25 @@ def test_len(random_dataset: BaseDataset):
     assert len(random_dataset) == 5  # last dimension is 5 volumes
 
 
-def test_getitem_volume_index(random_dataset: BaseDataset[()]):
+def test_getitem_volume_index(random_dataset: BaseDataset):
     """
     Test that __getitem__ returns the correct (volume, affine) tuple.
 
     By default, motion_affines is None, so we expect to get None for the affine.
     """
-    # Single volume
-    volume0, aff0 = random_dataset[0]
+    # Single volume  # Note that the type ignore can be removed once we can use *Ts
+    volume0, aff0 = random_dataset[0]  # type: ignore[misc]  # PY310
     assert volume0.shape == (32, 32, 32)
     # No transforms have been applied yet, so there's no motion_affines array
     assert aff0 is None
 
     # Slice of volumes
-    volume_slice, aff_slice = random_dataset[2:4]
+    volume_slice, aff_slice = random_dataset[2:4]  # type: ignore[misc]  # PY310
     assert volume_slice.shape == (32, 32, 32, 2)
     assert aff_slice is None
 
 
-def test_set_transform(random_dataset: BaseDataset[()]):
+def test_set_transform(random_dataset: BaseDataset):
     """
     Test that calling set_transform changes the data and motion_affines.
     For simplicity, we'll apply an identity transform and check that motion_affines is updated.
@@ -83,7 +83,7 @@ def test_set_transform(random_dataset: BaseDataset[()]):
     random_dataset.set_transform(idx, affine, order=1)
 
     # Data shouldn't have changed (since transform is identity).
-    volume0, aff0 = random_dataset[idx]
+    volume0, aff0 = random_dataset[idx]  # type: ignore[misc]  # PY310
     assert np.allclose(data_before, volume0)
 
     # motion_affines should be created and match the transform matrix.

--- a/tox.ini
+++ b/tox.ini
@@ -2,13 +2,17 @@
 requires =
   tox>=4
 envlist =
-  py312, notebooks
+  py3{10,11,12,13}
+  notebooks
 skip_missing_interpreters = true
 
 # Configuration that allows us to split tests across GitHub runners effectively
 [gh-actions]
 python =
+  3.10: py310
+  3.11: py311
   3.12: py312, notebooks
+  3.13: py313
 
 [testenv]
 description = Pytest with coverage

